src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -132,14 +133,21 @@ private void Dispose(bool disposing)
                     GC.SuppressFinalize(this);
                     _stream.Dispose();
 
-                    // Eat any exceptions from the read-ahead task.  We don't need to log, as we expect
+                    // Eat any exceptions from the read-ahead task. We don't need to log, as we expect
                     // failures from this task due to closing the connection while a read is in progress.
                     ValueTask<int>? readAheadTask = ConsumeReadAheadTask();
                     if (readAheadTask != null)
                     {
                         IgnoreExceptions(readAheadTask.GetValueOrDefault());
                     }
                 }
+                else
+                {
+                    Debug.Assert(_readAheadTask is not ValueTask<int> t
+                        || t.IsCompletedSuccessfully
+                        || t.AsTask().GetAwaiter().GetResult() >= 0, // This may throw on the finalizer thread
+                        $"The readAheadTask threw an exception that was not caught, or ignored during disposal: {_readAheadTask!.Value.AsTask().Exception}");
+                }
```

Comment:
Rather than storing the exception into a field, why can't we just remove the `if (disposing)` here?

Reply:
> Rather than storing the exception into a field, why can't we just remove the if (disposing) here?

I was testing with this initially, but since `TaskExceptionHolder` in `_readAheadTask` is also finalizable, its `Finalize` may run before `HttpConnection`'s finalizer. Doing so would reduce the amount of unobserved exceptions, but wouldn't prevent them entirely.

Reply:
Ah, of course, i.e. the entire reason the bool is passed to Dispose. Ok.